plotly/plotly/plotly.py

```diff
@@ -80,68 +80,55 @@ def get_config():
 
+        """
+        updated_plot_options = {}
+        base_plot_options = copy.deepcopy(base_plot_options)
+        new_plot_options = copy.deepcopy(new_plot_options)
+        for key in base_plot_options:
+            updated_plot_options[key] = new_plot_options.get(
+                key, base_plot_options[key])
 
-def _plot_option_logic(plot_options):
-    """
-    Given some plot_options as part of a plot call, decide on final options
+        return updated_plot_options
 
-    """
-    session_plot_options = get_session_plot_options()
-    current_plot_options = get_plot_options()
-    current_plot_options.update(plot_options)
-    if (('filename' in plot_options or 'filename' in session_plot_options) and
-            'fileopt' not in session_plot_options and
-            'fileopt' not in plot_options):
-        current_plot_options['fileopt'] = 'overwrite'
-
-    # Check for any conflicts between 'sharing' and 'world_readable'
-    if 'sharing' in plot_options:
-        if plot_options['sharing'] in ['public', 'private', 'secret']:
-
-            if 'world_readable' not in plot_options:
-                if plot_options['sharing'] != 'public':
-                    current_plot_options['world_readable'] = False
-                else:
-                    current_plot_options['world_readable'] = True
-            elif (plot_options['world_readable'] and
-                    plot_options['sharing'] != 'public'):
-                raise exceptions.PlotlyError(
-                    "Looks like you are setting your plot privacy to both "
-                    "public and private.\n If you set world_readable as True, "
-                    "sharing can only be set to 'public'")
-            elif (not plot_options['world_readable'] and
-                    plot_options['sharing'] == 'public'):
-                raise exceptions.PlotlyError(
-                    "Looks like you are setting your plot privacy to both "
-                    "public and private.\n If you set world_readable as "
-                    "False, sharing can only be set to 'private' or 'secret'")
-        else:
-            raise exceptions.PlotlyError(
-                "The 'sharing' argument only accepts one of the following "
-                "strings:\n'public' -- for public plots\n"
-                "'private' -- for private plots\n"
-                "'secret' -- for private plots that can be shared with a "
-                "secret url"
-            )
-    else:
-        current_plot_options['sharing'] = None
+    user_plot_options = default_plot_options
+    user_plot_options = update_plot_options(user_plot_options,
+                                            file_options)
+    user_plot_options = update_plot_options(user_plot_options,
+                                            session_options)
+    user_plot_options = update_plot_options(user_plot_options,
+                                            plot_options_from_call_signature)
```

Comment:
:cow2: Sorry, I still don't understand why we need to do the `update_plot_options` thing? It just seems like it'd be clearer to strip them out later?

```
user_plot_options = {}
user_plot_options.update(default_plot_options)
user_plot_options.update(file_options)
user_plot_options.update(session_options)
user_plot_options.update(plot_optoins_from_call_signature)
user_plot_options = {k: v for k, v in user_plot_options.items()
                     if k in default_plot_options}
```

It's basically the same number of lines and doesn't introduce anything new.